tensorboard/webapp/plugins/npmi/views/violin_filters/violin_filter/violin_filter_component.ts

```diff
@@ -0,0 +1,413 @@
+      .attr('transform', `translate(${this.margin.left}, ${this.margin.top})`);
+    this.drawContainer = this.mainContainer
+      .append('g')
+      .attr(
+        'transform',
+        `translate(${this.drawMargin.left}, ${this.drawMargin.top})`
+      );
+    this.dotsGroup = this.drawContainer.append('g').attr('class', 'dotsGroup');
+    this.yAxisGroup = this.mainContainer
+      .append('g')
+      .attr('class', 'axis axis--y');
+    this.xAxisGroup = this.mainContainer
+      .append('g')
+      .attr('class', 'axis axis--x');
+    this.miscGroup = this.drawContainer.append('g');
+    this.initializeBrush();
+    this.drawMisc();
+    this.redraw();
+  }
+
+  ngOnChanges(changes: SimpleChanges) {
+    if (this.svg) {
+      this.redraw();
+    }
+  }
+
+  redraw() {
+    this.setMaxBinSize();
+    this.updateAxes();
+    this.draw();
+  }
+
+  // Initializing/Updating the visualization props
+  setMaxBinSize() {
+    Object.values(this.chartData.violinData).forEach((dataElement) => {
+      const lengths = dataElement.map((bin) => bin.length);
+      const longest: number = Math.max(...lengths);
+      this.maxBinSize = Math.max(longest, this.maxBinSize);
+    });
+  }
+
+  updateAxes() {
+    this.xScale = d3
+      .scaleBand()
+      .range([0, this.drawWidth])
+      .domain(Object.keys(this.chartData.violinData))
+      .padding(0.05);
+    this.xAxis = d3.axisBottom(this.xScale);
+
+    this.yScale = d3
+      .scaleLinear()
+      .range([this.drawHeight, 0])
+      .domain([this.chartData.extremes.min, this.chartData.extremes.max]);
+    this.yAxis = d3.axisLeft(this.yScale);
+
+    this.xScaleNum = d3
+      .scaleLinear()
+      .range([0, this.xScale.bandwidth()])
+      .domain([-this.maxBinSize, this.maxBinSize]);
+  }
+
+  initializeBrush() {
+    this.brush = d3
+      .brushY()
+      .extent([
+        [
+          this.margin.left + this.drawMargin.left,
+          this.margin.top + this.drawMargin.top,
+        ],
+        [
+          this.width - this.margin.right,
+          this.height - this.margin.bottom / 2.0,
+        ],
+      ])
+      .on('end', this.brushMoved.bind(this));
+  }
+
+  // Drawing UI
+  draw() {
+    this.drawAxes();
+    this.drawPlot();
+    this.refreshMisc();
+    this.refreshBrush();
+  }
+
+  drawAxes() {
+    this.miscGroup.selectAll('*').remove();
+    this.miscGroup
+      .append('line')
+      .style('stroke', 'black')
+      .attr('x1', 0)
+      .attr('y1', this.yScale(0))
+      .attr('x2', this.drawWidth)
+      .attr('y2', this.yScale(0));
+    this.yAxisGroup
+      .attr(
+        'transform',
+        `translate(${this.drawMargin.left},
+      ${this.drawMargin.top})`
+      )
+      .call(this.yAxis);
+    this.xAxisGroup
+      .attr(
+        'transform',
+        `translate(${this.drawMargin.left},
+      ${this.drawMargin.top + this.chartHeight})`
+      )
+      .call(this.xAxis);
+    this.miscGroup
+      .append('text')
+      .style('fill', 'black')
```

Comment:
Just wondering, would it be better to style these nodes using CSS instead?

Reply:
I think there are different ways to look at it. Like this, you have all the attributes needed to understand how the element looks in one place. Additionally, in most cases, this would mean changing one `.style(` for a `.class(`. On the other hand, if you want to modify the style of the visualizations, you need to change the code and cant just exchange a CSS file. If it were moved to CSS, however, I would need to modify ViewEncapsulation for this.

All in all, I think keeping it with the component for the visualizations is the better way, as basically all d3 code is determining the looks of a dom element, so moving out the `.style` of it would not make a huge improvement IMO compared to its downsides.

Happy to change this, if you feel otherwise, however. Let me know what you think.